kevlar/include/kevlar_bits/model/binomial/berry_inla.hpp

```diff
@@ -0,0 +1,368 @@
+          class _GridRangeType>
+struct BerryINLA<ValueType, ARMS>::SimGlobalState
+    : SimGlobalStateFixedNDefault<_GenType, _ValueType, _UIntType,
+                                  _GridRangeType> {
+    struct SimState;
+
+    using base_t = SimGlobalStateFixedNDefault<_GenType, _ValueType, _UIntType,
+                                               _GridRangeType>;
+    using typename base_t::gen_t;
+    using typename base_t::grid_range_t;
+    using typename base_t::interface_t;
+    using typename base_t::uint_t;
+    using typename base_t::value_t;
+
+    using sim_state_t = SimState;
+
+   private:
+    using model_t = BerryINLA;
+    const model_t& model;
+
+   public:
+    SimGlobalState(const model_t& model, const grid_range_t& grid_range)
+        : base_t(model.n_arm_samples, grid_range), model(model) {}
+
+    std::unique_ptr<typename interface_t::sim_state_t> make_sim_state(
+        size_t seed) const override {
+        return std::make_unique<sim_state_t>(*this, seed);
+    }
+};
+
+template <class ValueType, int ARMS>
+template <class _GenType, class _ValueType, class _UIntType,
+          class _GridRangeType>
+struct BerryINLA<ValueType, ARMS>::SimGlobalState<
+    _GenType, _ValueType, _UIntType, _GridRangeType>::SimState
+    : base_t::sim_state_t {
+   private:
+    using outer_t = SimGlobalState;
+
+   public:
+    using base_t = typename outer_t::base_t::sim_state_t;
+    using typename base_t::interface_t;
+
+   private:
+    const outer_t& outer_;
+
+   public:
+    SimState(const outer_t& sgs, size_t seed)
+        : base_t(sgs, seed), outer_(sgs) {}
+
+    void simulate(Eigen::Ref<colvec_type<uint_t>> rej_len) override {
+        base_t::generate_data();
+        base_t::generate_sufficient_stats();
+
+        const auto& bits = outer_.bits();
+        const auto& gr = outer_.grid_range();
+
+        assert(gr.n_params() == ARMS);
+
+        const auto n_arm_samples = outer_.model.n_arm_samples;
+        const auto& critical_values = outer_.model.critical_values();
+
+        size_t pos = 0;
+        for (size_t grid_i = 0; grid_i < gr.n_gridpts(); ++grid_i) {
+            auto bits_i = bits.col(grid_i);
+
+            vec_t y(ARMS);
+            for (int i = 0; i < y.size(); ++i) {
+                const auto& ss_i = base_t::sufficient_stats_arm(i);
+                y(i) = static_cast<value_t>(ss_i(bits_i[i]));
+            }
+
+            vec_t exceedance = outer_.model.get_posterior_exceedance_probs(y);
+
+            for (size_t n_t = 0; n_t < gr.n_tiles(grid_i); ++n_t, ++pos) {
+                value_t max_null_prob_exceed = 0;
+                for (int arm_i = 0; arm_i < ARMS; ++arm_i) {
+                    if (gr.check_null(pos, arm_i)) {
+                        max_null_prob_exceed =
+                            std::max(max_null_prob_exceed, exceedance[arm_i]);
+                    }
+                }
+
+                int cv_i = 0;
+                for (; cv_i < critical_values.size(); ++cv_i) {
```

Comment:
std::lower_bound might save ~5 lines here

Reply:
I would strongly strongly discourage use of stl algorithms. I saved so much freakin time just by writing my own loops instead of std::find_if for this exact part. Keep in mind the critical values I was testing with were length 1 so it's not even a big O complexity difference. Please please benchmark if you're gonna make that change.

Reply:
I don't think I'm going to change it!

Reply:
Fair. This is a vector of floats so I would hope the STL would be foolproof. At any rate, assuming the profile is similar to dirty bayes, this block is very cold.

Reply:
Yeah, that block is cold as ice. hhaha. I'm mostly not changing for the same reason I've commented on other parts of the C++ code. "I'd actually lean towards removing this eventually and using James' new trampoline/Python model stuff. The INLA stuff is just so much prettier written in JAX/numpy."

Reply:
Very much looking forward to the Python version